### ISSUE TITLE
NaN when computing gradient of squared norm evaluated at 0.

### ISSUE THREAD

**Issue by ghimiremukesh** [google/jax#21723]
### Description

`jax.grad` returns `nan` when computing the gradient of a norm evaluated at 0. 

Following code returns `nan`
`import jax`
`import jax.numpy as jnp`

`f = lambda x: jnp.linalg.norm(x) ** 2`
`jax.grad(f)(0.)`

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.4.26
jaxlib: 0.4.26
numpy:  1.26.4
python: 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:48:25)  [Clang 14.0.6 ]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='EN4147604-SEMTE-2.local', release='23.5.0', version='Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000', machine='arm64')

**Comment by jakevdp** [comment]
Hi, thanks for the report! This is working as intended: the gradient of the norm at 0.0 is not well-defined, so NaN is the most reasonable output.

Mathematically, you have

$norm(x) = \left[\sum_i x_i^2\right]^{1/2}$

The partial derivative with respect to each input is

$d/dx_i norm(x) = x_i  \left[\sum_i x_i^2\right]^{-1/2} = x_i/norm(x)$

And if you evaluate this at $x=0$ you get $0/0$, which is undefined.

**Comment by ghimiremukesh** [comment]
That is true. For squared norm however, it would be $2x_i$. I can instead use jnp.sum(x**2) to avoid the issue. Thanks!
